pkg/restic/repository_manager.go

```diff
@@ -0,0 +1,404 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restic
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	corev1api "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kerrs "k8s.io/apimachinery/pkg/util/errors"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	arkv1api "github.com/heptio/ark/pkg/apis/ark/v1"
+	"github.com/heptio/ark/pkg/cloudprovider"
+)
+
+// TODO this is more like a metadata manager
+type RepositoryManager interface {
+	RepoPrefix() string
+	RepositoryExists(name string) (bool, error)
+	InitRepo(name string) error
+	CheckRepo(name string) error
+	CheckAllRepos() error
+	PruneRepo(name string) error
+	PruneAllRepos() error
+
+	GetSnapshotID(repo, backupUID, podUID, volume string) (string, error)
+	Forget(repo, snapshotID string) error
+}
+
+type repositoryManager struct {
+	objectStore   cloudprovider.ObjectStore
+	bucket        string
+	secretsClient corev1client.SecretInterface
+	log           logrus.FieldLogger
+	repoPrefix    string
+}
+
+type BackendType string
+
+const (
+	AWSBackend   BackendType = "aws"
+	AzureBackend BackendType = "azure"
+	GCPBackend   BackendType = "gcp"
+)
+
+const (
+	credsSecret   = "restic-credentials"
+	credsFilePath = "/restic-credentials/%s"
+)
+
+// NewRepositoryManager constructs a RepositoryManager.
+func NewRepositoryManager(objectStore cloudprovider.ObjectStore, config arkv1api.ObjectStorageProviderConfig, secretsClient corev1client.SecretInterface, log logrus.FieldLogger) RepositoryManager {
+	rm := &repositoryManager{
+		objectStore:   objectStore,
+		bucket:        config.ResticLocation,
+		secretsClient: secretsClient,
+		log:           log,
+	}
+
+	switch BackendType(config.Name) {
+	case AWSBackend:
+		url := "s3.amazonaws.com"
+
+		// non-AWS, S3-compatible object store
+		if config.Config != nil && config.Config["s3Url"] != "" {
+			url = config.Config["s3Url"]
+		}
+
+		rm.repoPrefix = fmt.Sprintf("s3:%s/%s", url, rm.bucket)
+	case AzureBackend:
+		rm.repoPrefix = "azure:" + rm.bucket + ":"
+	case GCPBackend:
+		rm.repoPrefix = "gs:" + rm.bucket + ":"
+	}
+
+	return rm
+}
+
+func (rm *repositoryManager) RepoPrefix() string {
+	return rm.repoPrefix
+}
+
+func (rm *repositoryManager) RepositoryExists(name string) (bool, error) {
+	repos, err := rm.getAllRepos()
+	if err != nil {
+		return false, err
+	}
+
+	for _, repo := range repos {
+		if repo == name {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (rm *repositoryManager) InitRepo(name string) error {
+	resticCreds, err := rm.secretsClient.Get(credsSecret, metav1.GetOptions{})
```

Comment:
The plan was to require an admin to precreate manually per namespace, wasn't it?

Reply:
We should discuss the ui/ux. We could have something like
```
# generates an encryption key, creates a secret for it, maybe creates a new ResticRepo, ark server sees it's New and runs restic init?
ark restic init --namespace foo

# same as above, but uses a precreated encryption key instead of generating
ark restic init --namespace foo --encryption-key /path/to/file
```

Reply:
If we were to add a ResticRepo CRD, we could use that to manage its status (needs init, already done, etc). Not sure if we need this or how it would impact backup replication. Would probably have to be tied to a BackupTarget when we get there.